File: backend/app/utils/retry_utils.py
"""Retry and error recovery utilities for backend services."""

import asyncio
import logging
import random
import time
from datetime import datetime
from enum import Enum
from functools import wraps
from typing import Any, Callable, Dict, List, Optional, Type

from app.exceptions import (
    AIServiceError,
    BaseAppException,
    DatabaseError,
    ErrorCategory,
    ExternalAPIError,
)

logger = logging.getLogger(__name__)


class CircuitBreakerState(str, Enum):
    """Circuit breaker states."""

    CLOSED = "closed"
    OPEN = "open"
    HALF_OPEN = "half_open"


class RetryConfig:
    """Configuration for retry mechanisms."""

    def __init__(
        self,
        max_attempts: int = 3,
        base_delay: float = 1.0,
        max_delay: float = 60.0,
        exponential_backoff: bool = True,
        jitter: bool = True,
        retry_exceptions: Optional[List[Type[Exception]]] = None,
        stop_exceptions: Optional[List[Type[Exception]]] = None,
    ):
        self.max_attempts = max_attempts
        self.base_delay = base_delay
        self.max_delay = max_delay
        self.exponential_backoff = exponential_backoff
        self.jitter = jitter
        self.retry_exceptions = retry_exceptions or []
        self.stop_exceptions = stop_exceptions or []


def get_default_retry_config(category: ErrorCategory) -> RetryConfig:
    """Get default retry configuration for different error categories."""

    if category == ErrorCategory.AI_SERVICE:
        return RetryConfig(
            max_attempts=3,
            base_delay=2.0,
            max_delay=30.0,
            retry_exceptions=[AIServiceError, ConnectionError, TimeoutError],
            stop_exceptions=[ValueError, TypeError],
        )

    elif category == ErrorCategory.DATABASE:
        return RetryConfig(
            max_attempts=2,
            base_delay=1.0,
            max_delay=10.0,
            retry_exceptions=[DatabaseError, ConnectionError],
            stop_exceptions=[ValueError],
        )

    elif category == ErrorCategory.EXTERNAL_API:
        return RetryConfig(
            max_attempts=5,
            base_delay=0.5,
            max_delay=15.0,
            retry_exceptions=[ExternalAPIError, ConnectionError, TimeoutError],
        )

    else:
        return RetryConfig(
            max_attempts=3,
            base_delay=1.0,
            max_delay=10.0,
        )


def calculate_delay(
    attempt: int,
    base_delay: float,
    max_delay: float,
    exponential_backoff: bool = True,
    jitter: bool = True,
) -> float:
    """Calculate delay with exponential backoff and jitter."""

    if exponential_backoff:
        delay = base_delay * (2 ** (attempt - 1))
    else:
        delay = base_delay

    # Apply maximum delay limit
    delay = min(delay, max_delay)

    # Add jitter to prevent thundering herd
    if jitter:
        delay = delay * (0.5 + random.random() * 0.5)

    return delay


def should_retry_exception(
    exception: Exception,
    retry_exceptions: List[Type[Exception]],
    stop_exceptions: List[Type[Exception]],
) -> bool:
    """Determine if an exception should trigger a retry."""

    # Never retry stop exceptions
    for stop_exc in stop_exceptions:
        if isinstance(exception, stop_exc):
            return False

    # If retry exceptions are specified, only retry those
    if retry_exceptions:
        for retry_exc in retry_exceptions:
            if isinstance(exception, retry_exc):
                return True
        return False

    # Default: retry BaseAppException with retryable=True
    if isinstance(exception, BaseAppException):
        return exception.retryable

    return False


async def retry_async(
    func: Callable,
    config: Optional[RetryConfig] = None,
    context: Optional[Dict[str, Any]] = None,
    *args,
    **kwargs,
) -> Any:
    """Retry an async function with configurable backoff."""

    if config is None:
        config = RetryConfig()

    context = context or {}
    last_exception = None

    for attempt in range(1, config.max_attempts + 1):
        try:
            result = await func(*args, **kwargs)

            if attempt > 1:
                logger.info(
                    f"Function {func.__name__} succeeded after {attempt} attempts",
                    extra={
                        "function": func.__name__,
                        "attempt": attempt,
                        "max_attempts": config.max_attempts,
                        **context,
                    },
                )

            return result

        except Exception as e:
            last_exception = e

            # Check if we should retry this exception
            if not should_retry_exception(
                e, config.retry_exceptions, config.stop_exceptions
            ):
                logger.warning(
                    f"Exception {type(e).__name__} is not retryable, failing immediately",
                    extra={
                        "function": func.__name__,
                        "exception": str(e),
                        "attempt": attempt,
                        **context,
                    },
                )
                raise

            # If this was the last attempt, don't retry
            if attempt == config.max_attempts:
                logger.error(
                    f"Function {func.__name__} failed after {config.max_attempts} attempts",
                    extra={
                        "function": func.__name__,
                        "exception": str(e),
                        "attempt": attempt,
                        "max_attempts": config.max_attempts,
                        **context,
                    },
                )
                raise

            # Calculate delay and wait
            delay = calculate_delay(
                attempt,
                config.base_delay,
                config.max_delay,
                config.exponential_backoff,
                config.jitter,
            )

            logger.warning(
                f"Function {func.__name__} failed on attempt {attempt}, retrying in {delay:.2f}s",
                extra={
                    "function": func.__name__,
                    "exception": str(e),
                    "attempt": attempt,
                    "max_attempts": config.max_attempts,
                    "delay": delay,
                    **context,
                },
            )

            await asyncio.sleep(delay)

    # This should never be reached, but just in case
    raise last_exception


def retry_sync(
    func: Callable,
    config: Optional[RetryConfig] = None,
    context: Optional[Dict[str, Any]] = None,
    *args,
    **kwargs,
) -> Any:
    """Retry a sync function with configurable backoff."""

    if config is None:
        config = RetryConfig()

    context = context or {}
    last_exception = None

    for attempt in range(1, config.max_attempts + 1):
        try:
            result = func(*args, **kwargs)

            if attempt > 1:
                logger.info(
                    f"Function {func.__name__} succeeded after {attempt} attempts",
                    extra={
                        "function": func.__name__,
                        "attempt": attempt,
                        "max_attempts": config.max_attempts,
                        **context,
                    },
                )

            return result

        except Exception as e:
            last_exception = e

            # Check if we should retry this exception
            if not should_retry_exception(
                e, config.retry_exceptions, config.stop_exceptions
            ):
                logger.warning(
                    f"Exception {type(e).__name__} is not retryable, failing immediately",
                    extra={
                        "function": func.__name__,
                        "exception": str(e),
                        "attempt": attempt,
                        **context,
                    },
                )
                raise

            # If this was the last attempt, don't retry
            if attempt == config.max_attempts:
                logger.error(
                    f"Function {func.__name__} failed after {config.max_attempts} attempts",
                    extra={
                        "function": func.__name__,
                        "exception": str(e),
                        "attempt": attempt,
                        "max_attempts": config.max_attempts,
                        **context,
                    },
                )
                raise

            # Calculate delay and wait
            delay = calculate_delay(
                attempt,
                config.base_delay,
                config.max_delay,
                config.exponential_backoff,
                config.jitter,
            )

            logger.warning(
                f"Function {func.__name__} failed on attempt {attempt}, retrying in {delay:.2f}s",
                extra={
                    "function": func.__name__,
                    "exception": str(e),
                    "attempt": attempt,
                    "max_attempts": config.max_attempts,
                    "delay": delay,
                    **context,
                },
            )

            time.sleep(delay)

    # This should never be reached, but just in case
    raise last_exception


def with_retry(
    config: Optional[RetryConfig] = None, context: Optional[Dict[str, Any]] = None
):
    """Decorator for adding retry logic to functions."""

    def decorator(func: Callable) -> Callable:
        if asyncio.iscoroutinefunction(func):

            @wraps(func)
            async def async_wrapper(*args, **kwargs):
                return await retry_async(func, config, context, *args, **kwargs)

            return async_wrapper
        else:

            @wraps(func)
            def sync_wrapper(*args, **kwargs):
                return retry_sync(func, config, context, *args, **kwargs)

            return sync_wrapper

    return decorator


class CircuitBreaker:
    """Circuit breaker implementation for preventing cascading failures."""

    def __init__(
        self,
        failure_threshold: int = 5,
        reset_timeout: float = 60.0,
        expected_exception: Type[Exception] = Exception,
        name: str = "CircuitBreaker",
    ):
        self.failure_threshold = failure_threshold
        self.reset_timeout = reset_timeout
        self.expected_exception = expected_exception
        self.name = name

        self.failure_count = 0
        self.last_failure_time: Optional[datetime] = None
        self.state = CircuitBreakerState.CLOSED
        self.success_count = 0

    def _should_attempt_reset(self) -> bool:
        """Check if circuit breaker should attempt to reset."""
        if self.last_failure_time is None:
            return False

        time_since_failure = datetime.utcnow() - self.last_failure_time
        return time_since_failure.total_seconds() >= self.reset_timeout

    def _on_success(self) -> None:
        """Handle successful operation."""
        self.failure_count = 0

        if self.state == CircuitBreakerState.HALF_OPEN:
            self.success_count += 1
            if self.success_count >= 3:  # Require 3 successes to fully close
                self.state = CircuitBreakerState.CLOSED
                self.success_count = 0
                logger.info(f"Circuit breaker {self.name} transitioned to CLOSED")

    def _on_failure(self, exception: Exception) -> None:
        """Handle failed operation."""
        if not isinstance(exception, self.expected_exception):
            return

        self.failure_count += 1
        self.last_failure_time = datetime.utcnow()

        if self.state == CircuitBreakerState.HALF_OPEN:
            self.state = CircuitBreakerState.OPEN
            self.success_count = 0
            logger.warning(
                f"Circuit breaker {self.name} transitioned to OPEN from HALF_OPEN"
            )
        elif self.failure_count >= self.failure_threshold:
            self.state = CircuitBreakerState.OPEN
            logger.warning(
                f"Circuit breaker {self.name} transitioned to OPEN - failure threshold reached"
            )

    async def call_async(self, func: Callable, *args, **kwargs) -> Any:
        """Execute async function with circuit breaker protection."""

        if self.state == CircuitBreakerState.OPEN:
            if self._should_attempt_reset():
                self.state = CircuitBreakerState.HALF_OPEN
                logger.info(f"Circuit breaker {self.name} transitioned to HALF_OPEN")
            else:
                raise Exception(f"Circuit breaker {self.name} is OPEN")

        try:
            result = await func(*args, **kwargs)
            self._on_success()
            return result
        except Exception as e:
            self._on_failure(e)
            raise

    def call_sync(self, func: Callable, *args, **kwargs) -> Any:
        """Execute sync function with circuit breaker protection."""

        if self.state == CircuitBreakerState.OPEN:
            if self._should_attempt_reset():
                self.state = CircuitBreakerState.HALF_OPEN
                logger.info(f"Circuit breaker {self.name} transitioned to HALF_OPEN")
            else:
                raise Exception(f"Circuit breaker {self.name} is OPEN")

        try:
            result = func(*args, **kwargs)
            self._on_success()
            return result
        except Exception as e:
            self._on_failure(e)
            raise

    def get_state(self) -> CircuitBreakerState:
        """Get current circuit breaker state."""
        return self.state

    def get_metrics(self) -> Dict[str, Any]:
        """Get circuit breaker metrics."""
        return {
            "name": self.name,
            "state": self.state.value,
            "failure_count": self.failure_count,
            "success_count": self.success_count,
            "last_failure_time": (
                self.last_failure_time.isoformat() if self.last_failure_time else None
            ),
            "failure_threshold": self.failure_threshold,
            "reset_timeout": self.reset_timeout,
        }

    def reset(self) -> None:
        """Manually reset circuit breaker."""
        self.state = CircuitBreakerState.CLOSED
        self.failure_count = 0
        self.success_count = 0
        self.last_failure_time = None
        logger.info(f"Circuit breaker {self.name} manually reset")


def with_circuit_breaker(
    circuit_breaker: CircuitBreaker,
    context: Optional[Dict[str, Any]] = None,
):
    """Decorator for adding circuit breaker protection to functions."""

    def decorator(func: Callable) -> Callable:
        if asyncio.iscoroutinefunction(func):

            @wraps(func)
            async def async_wrapper(*args, **kwargs):
                return await circuit_breaker.call_async(func, *args, **kwargs)

            return async_wrapper
        else:

            @wraps(func)
            def sync_wrapper(*args, **kwargs):
                return circuit_breaker.call_sync(func, *args, **kwargs)

            return sync_wrapper

    return decorator


# Global circuit breakers for common services
ai_service_circuit_breaker = CircuitBreaker(
    failure_threshold=3,
    reset_timeout=120.0,
    expected_exception=AIServiceError,
    name="AIService",
)

database_circuit_breaker = CircuitBreaker(
    failure_threshold=5,
    reset_timeout=60.0,
    expected_exception=DatabaseError,
    name="Database",
)

external_api_circuit_breaker = CircuitBreaker(
    failure_threshold=5,
    reset_timeout=30.0,
    expected_exception=ExternalAPIError,
    name="ExternalAPI",
)

File: frontend/src/utils/retryUtils.ts
/**
 * Retry utilities for error recovery mechanisms
 */

import { ErrorCategory } from "../types/error";
import { logWarn, logError, logInfo } from "./logger";

export interface RetryOptions {
  maxAttempts: number;
  baseDelay: number;
  maxDelay: number;
  exponentialBackoff: boolean;
  jitter: boolean;
  retryCondition?: (error: unknown) => boolean;
  onRetry?: (attempt: number, error: unknown) => void;
  onMaxAttemptsReached?: (error: unknown) => void;
}

export interface CircuitBreakerOptions {
  failureThreshold: number;
  resetTimeout: number;
  monitoringPeriod: number;
}

export enum CircuitBreakerState {
  CLOSED = "closed",
  OPEN = "open",
  HALF_OPEN = "half_open",
}

/**
 * Default retry options for different error categories
 */
export const getDefaultRetryOptions = (
  category: ErrorCategory,
): RetryOptions => {
  const baseOptions: RetryOptions = {
    maxAttempts: 3,
    baseDelay: 1000,
    maxDelay: 10000,
    exponentialBackoff: true,
    jitter: true,
  };

  switch (category) {
    case ErrorCategory.NETWORK:
      return {
        ...baseOptions,
        maxAttempts: 5,
        baseDelay: 500,
        retryCondition: (error) => {
          const status = (error as { response?: { status?: number } }).response
            ?.status;
          return !status || status >= 500 || status === 408 || status === 429;
        },
      };

    case ErrorCategory.AI_SERVICE:
      return {
        ...baseOptions,
        maxAttempts: 3,
        baseDelay: 2000,
        maxDelay: 30000,
        retryCondition: (error) => {
          const status = (error as { response?: { status?: number } }).response
            ?.status;
          return status === 503 || status === 502 || status === 429;
        },
      };

    case ErrorCategory.API:
      return {
        ...baseOptions,
        maxAttempts: 3,
        baseDelay: 1000,
        retryCondition: (error) => {
          const status = (error as { response?: { status?: number } }).response
            ?.status;
          return status >= 500 || status === 429;
        },
      };

    case ErrorCategory.DATABASE:
      return {
        ...baseOptions,
        maxAttempts: 2,
        baseDelay: 2000,
        retryCondition: (error) => {
          const status = (error as { response?: { status?: number } }).response
            ?.status;
          return status === 503 || status === 502;
        },
      };

    default:
      return {
        ...baseOptions,
        maxAttempts: 1,
        retryCondition: () => false,
      };
  }
};

/**
 * Calculate delay with exponential backoff and jitter
 */
const calculateDelay = (
  attempt: number,
  baseDelay: number,
  maxDelay: number,
  exponentialBackoff: boolean,
  jitter: boolean,
): number => {
  let delay = exponentialBackoff
    ? baseDelay * Math.pow(2, attempt - 1)
    : baseDelay;

  // Apply maximum delay limit
  delay = Math.min(delay, maxDelay);

  // Add jitter to prevent thundering herd
  if (jitter) {
    delay = delay * (0.5 + Math.random() * 0.5);
  }

  return Math.floor(delay);
};

/**
 * Sleep utility for delays
 */
const sleep = (ms: number): Promise<void> => {
  return new Promise((resolve) => setTimeout(resolve, ms));
};

/**
 * Retry function with configurable options
 */
export async function retryWithBackoff<T>(
  operation: () => Promise<T>,
  options: Partial<RetryOptions> = {},
  context?: { component?: string; action?: string },
): Promise<T> {
  const config: RetryOptions = {
    maxAttempts: 3,
    baseDelay: 1000,
    maxDelay: 10000,
    exponentialBackoff: true,
    jitter: true,
    retryCondition: () => true,
    ...options,
  };

  let lastError: unknown;

  for (let attempt = 1; attempt <= config.maxAttempts; attempt++) {
    try {
      const result = await operation();

      if (attempt > 1) {
        logInfo(`Operation succeeded after ${attempt} attempts`, {
          component: context?.component,
          action: context?.action,
          attempt,
          maxAttempts: config.maxAttempts,
        });
      }

      return result;
    } catch (error) {
      lastError = error;

      // Check if we should retry this error
      if (!config.retryCondition!(error)) {
        logWarn("Error not retryable, failing immediately", {
          component: context?.component,
          action: context?.action,
          error: (error as Error).message,
          attempt,
        });
        throw error;
      }

      // If this was the last attempt, don't retry
      if (attempt === config.maxAttempts) {
        logError("Max retry attempts reached", {
          component: context?.component,
          action: context?.action,
          error: (error as Error).message,
          attempt,
          maxAttempts: config.maxAttempts,
        });

        config.onMaxAttemptsReached?.(error);
        throw error;
      }

      // Calculate delay and wait
      const delay = calculateDelay(
        attempt,
        config.baseDelay,
        config.maxDelay,
        config.exponentialBackoff,
        config.jitter,
      );

      logWarn(`Operation failed, retrying in ${delay}ms`, {
        component: context?.component,
        action: context?.action,
        error: (error as Error).message,
        attempt,
        maxAttempts: config.maxAttempts,
        delay,
      });

      config.onRetry?.(attempt, error);
      await sleep(delay);
    }
  }

  throw lastError;
}

/**
 * Circuit breaker implementation for preventing cascading failures
 */
export class CircuitBreaker {
  private state: CircuitBreakerState = CircuitBreakerState.CLOSED;
  private failureCount = 0;
  private lastFailureTime?: Date;
  private successCount = 0;

  constructor(
    private options: CircuitBreakerOptions,
    private name = "CircuitBreaker",
  ) {}

  async execute<T>(operation: () => Promise<T>): Promise<T> {
    if (this.state === CircuitBreakerState.OPEN) {
      if (this.shouldAttemptReset()) {
        this.state = CircuitBreakerState.HALF_OPEN;
        logInfo(`Circuit breaker ${this.name} transitioning to HALF_OPEN`);
      } else {
        const error = new Error(`Circuit breaker ${this.name} is OPEN`);
        logWarn(
          `Circuit breaker ${this.name} rejecting request - circuit is OPEN`,
        );
        throw error;
      }
    }

    try {
      const result = await operation();
      this.onSuccess();
      return result;
    } catch (error) {
      this.onFailure();
      throw error;
    }
  }

  private shouldAttemptReset(): boolean {
    if (!this.lastFailureTime) return false;

    const timeSinceLastFailure = Date.now() - this.lastFailureTime.getTime();
    return timeSinceLastFailure >= this.options.resetTimeout;
  }

  private onSuccess(): void {
    this.failureCount = 0;

    if (this.state === CircuitBreakerState.HALF_OPEN) {
      this.successCount++;
      if (this.successCount >= 3) {
        // Require 3 successes to fully close
        this.state = CircuitBreakerState.CLOSED;
        this.successCount = 0;
        logInfo(`Circuit breaker ${this.name} transitioning to CLOSED`);
      }
    }
  }

  private onFailure(): void {
    this.failureCount++;
    this.lastFailureTime = new Date();

    if (this.state === CircuitBreakerState.HALF_OPEN) {
      this.state = CircuitBreakerState.OPEN;
      this.successCount = 0;
      logWarn(
        `Circuit breaker ${this.name} transitioning to OPEN from HALF_OPEN`,
      );
    } else if (this.failureCount >= this.options.failureThreshold) {
      this.state = CircuitBreakerState.OPEN;
      logWarn(
        `Circuit breaker ${this.name} transitioning to OPEN - failure threshold reached`,
      );
    }
  }

  getState(): CircuitBreakerState {
    return this.state;
  }

  getMetrics() {
    return {
      state: this.state,
      failureCount: this.failureCount,
      successCount: this.successCount,
      lastFailureTime: this.lastFailureTime,
    };
  }

  reset(): void {
    this.state = CircuitBreakerState.CLOSED;
    this.failureCount = 0;
    this.successCount = 0;
    this.lastFailureTime = undefined;
    logInfo(`Circuit breaker ${this.name} manually reset`);
  }
}

/**
 * Automatic retry wrapper for API calls
 */
export function withRetry<T extends (...args: unknown[]) => Promise<unknown>>(
  fn: T,
  options: Partial<RetryOptions> = {},
  context?: { component?: string; action?: string },
): T {
  return (async (...args: Parameters<T>) => {
    return retryWithBackoff(() => fn(...args), options, context);
  }) as T;
}

/**
 * Bulk retry utility for multiple operations
 */
export async function retryBatch<T>(
  operations: Array<() => Promise<T>>,
  options: Partial<RetryOptions> = {},
): Promise<Array<{ success: boolean; result?: T; error?: unknown }>> {
  const results = await Promise.allSettled(
    operations.map((op) => retryWithBackoff(op, options)),
  );

  return results.map((result) => {
    if (result.status === "fulfilled") {
      return { success: true, result: result.value };
    } else {
      return { success: false, error: result.reason };
    }
  });
}

/**
 * Create a circuit breaker for API endpoints
 */
export function createApiCircuitBreaker(
  endpoint: string,
  options: Partial<CircuitBreakerOptions> = {},
): CircuitBreaker {
  const defaultOptions: CircuitBreakerOptions = {
    failureThreshold: 5,
    resetTimeout: 60000, // 1 minute
    monitoringPeriod: 10000, // 10 seconds
    ...options,
  };

  return new CircuitBreaker(defaultOptions, `API-${endpoint}`);
}

/**
 * Global circuit breakers for common services
 */
export const circuitBreakers = {
  api: createApiCircuitBreaker("api", {
    failureThreshold: 5,
    resetTimeout: 30000,
  }),
  workflows: createApiCircuitBreaker("workflows", {
    failureThreshold: 3,
    resetTimeout: 60000,
  }),
  ai: createApiCircuitBreaker("ai", {
    failureThreshold: 2,
    resetTimeout: 120000,
  }),
  templates: createApiCircuitBreaker("templates", {
    failureThreshold: 5,
    resetTimeout: 30000,
  }),
};


